File: src/utils/vol-regime.ts
/**
 * P3b — VIX-based volatility regime classifier.
 *
 * Source: docs/polymarket-prediction-improvements-research-2026-07.md §9
 *
 * Provides a synchronous, no-fitting alternative to running a secondary
 * HMM on (return, IV-change) joint dynamics. Uses VIX level as a regime
 * proxy and applies a per-step leverage-effect vol multiplier inside
 * the trajectory MC.
 *
 * Asset gating (§9.3): leverage effect applies only to **equity** and
 * **gold** — not to crypto (positive return-vol correlation in bull
 * markets) or commodities (regime-dependent).
 */

export type VolRegime = 'sticky_strike' | 'transitional' | 'sticky_implied_tree';

const _VIX_TRANSITIONAL = 15;
const _VIX_FEAR = 25;

/**
 * Classify the current vol regime from a VIX level.
 *
 *  - VIX < 15  → 'sticky_strike'
 *  - 15 ≤ VIX < 25 → 'transitional'
 *  - VIX ≥ 25  → 'sticky_implied_tree' (leverage-effect regime)
 */
export function getVolatilityRegime(vix: number): VolRegime {
  if (!Number.isFinite(vix) || vix < _VIX_TRANSITIONAL) return 'sticky_strike';
  if (vix < _VIX_FEAR) return 'transitional';
  return 'sticky_implied_tree';
}

/**
 * Per-step vol multiplier for the leverage effect in the trajectory MC.
 *
 *   regime=sticky_implied_tree, z<0 (down draw) → ×1.4 (vol amplifies)
 *   regime=sticky_implied_tree, z>0 (up draw)   → ×0.8 (vol mutes)
 *   else                                       → 1.0 (neutral)
 *
 * Gated on `assetClass ∈ {equity, gold}`. Other asset classes are
 * empirically not consistent with the leverage-effect pattern (§9.3).
 */
export function leverageVolMultiplier(
  regime: VolRegime,
  z: number,
  assetClass: string,
): number {
  if (regime !== 'sticky_implied_tree') return 1;
  if (assetClass !== 'equity' && assetClass !== 'gold') return 1;
  if (!Number.isFinite(z) || z === 0) return 1;
  return z < 0 ? 1.4 : 0.8;
}

File: research/models/vol_regime.py
"""P3b — VIX-based volatility regime classifier (Python mirror).

Source: docs/polymarket-prediction-improvements-research-2026-07.md §9

Mirrors ``src/utils/vol-regime.ts``.
"""

from __future__ import annotations

from math import isfinite
from typing import Literal

VolRegime = Literal["sticky_strike", "transitional", "sticky_implied_tree"]

_VIX_TRANSITIONAL = 15.0
_VIX_FEAR = 25.0


def get_volatility_regime(vix: float) -> VolRegime:
    if not isfinite(vix) or vix < _VIX_TRANSITIONAL:
        return "sticky_strike"
    if vix < _VIX_FEAR:
        return "transitional"
    return "sticky_implied_tree"


def leverage_vol_multiplier(regime: VolRegime, z: float, asset_class: str) -> float:
    """Per-step vol multiplier; gated on equity/gold (§9.3)."""
    if regime != "sticky_implied_tree":
        return 1.0
    if asset_class not in ("equity", "gold"):
        return 1.0
    if not isfinite(z) or z == 0:
        return 1.0
    return 1.4 if z < 0 else 0.8
